Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,534 @@
+        ts.tv_sec = self.nanoseconds / 1_000_000_000
+        ts.tv_nsec = self.nanoseconds % 1_000_000_000
+        return ts
+    }
+}
+
+// URingUserData supports (un)packing into an `UInt64` as io_uring has a user_data 64-bit payload which is set in the SQE
+// and returned in the CQE. We're using 56 of those 64 bits, 32 for the file descriptor, 16 for a "registration ID" and 8
+// for the type of event issued (poll/modify/delete).
+@usableFromInline struct URingUserData {
+    @usableFromInline var fileDescriptor: CInt
+    @usableFromInline var registrationID: UInt16 // SelectorRegistrationID truncated, only have room for bottom 16 bits (could be expanded to 24 if required)
+    @usableFromInline var eventType: CQEEventType
+    @usableFromInline var padding: Int8 // reserved for future use
+
+    @inlinable init(registrationID: SelectorRegistrationID, fileDescriptor: CInt, eventType: CQEEventType) {
+        assert(MemoryLayout<UInt64>.size == MemoryLayout<URingUserData>.size)
+        self.registrationID = UInt16(truncatingIfNeeded: registrationID.rawValue)
+        self.fileDescriptor = fileDescriptor
+        self.eventType = eventType
+        self.padding = 0
+    }
+
+    @inlinable init(rawValue: UInt64) {
+        let unpacked = IntegerBitPacking.unpackUInt32UInt16UInt8(rawValue)
+        self = .init(registrationID: SelectorRegistrationID(rawValue: UInt32(unpacked.1)),
+                     fileDescriptor: CInt(unpacked.0),
+                     eventType: CQEEventType(rawValue:unpacked.2)!)
+    }
+}
+
+extension UInt64 {
+    init(_ uringUserData: URingUserData) {
+        let fd = uringUserData.fileDescriptor
+        let eventType = uringUserData.eventType.rawValue
+        assert(fd >= 0, "\(fd) is not a valid file descriptor")
+        assert(eventType >= 0, "\(eventType) is not a valid eventType")
+
+        self = IntegerBitPacking.packUInt32UInt16UInt8(UInt32(truncatingIfNeeded: fd),
+                                                       uringUserData.registrationID,
+                                                       eventType)
+    }
+}
+
+// These are the events returned up to the selector
+internal struct UringEvent {
+    var fd : CInt
+    var pollMask : UInt32
+    var registrationID : UInt16 // we just have the truncated lower 16 bits of the registrationID
+    var pollCancelled : Bool
+    init () {
+        self.fd = -1
+        self.pollMask = 0
+        self.registrationID = 0
+        self.pollCancelled = false
+    }
+}
+
+// This is the key we use for merging events in our internal hashtable
+struct FDEventKey: Hashable {
+    var fileDescriptor : CInt
+    var registrationID : UInt16 // we just have the truncated lower 16 bits of the registrationID
+
+    init(_ f: CInt, _ s : UInt16) {
+        self.fileDescriptor = f
+        self.registrationID = s
+    }
+}
+
+final internal class Uring {
+    internal static let POLLIN: CUnsignedInt = numericCast(CNIOLinux.POLLIN)
```

Comment:
nits: usually I'd suggest to name these in a Swift way (ie. `pollIn`). Also `URing` vs `Uring`.

Reply:
The names were modelled after how its done in epoll-uring - I think it's nice if the style is consistent? (also for someone used to working with poll interfaces, they are immediately recognisable for what they are) - I have updated to URing though.